### PR TITLE
Ex3 adam

### PR DESCRIPTION
**Pull request by amonsonego**

### CODE REVIEW

File: app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt [hunk 1]
```diff
@@ -1,5 +1,15 @@
 package com.lightricks.feedexercise.network
 
-/**
- * todo: add Data Transfer Object data class(es) here
- */
+data class FeedMetadataDto(
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:7]: Let Moshi generates a small and fast adapter for each of your Kotlin classes at compile time.  Also you should annotate all the fields of the class with `Json`. This will prevent them from being removed/obfuscated from your code by R8/ProGuard.

File: app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+    val templatesMetadata: List<FeedItemDto>
+)
+
+data class FeedItemDto(
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:13]: Same comment.

File: app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt [hunk 3]
```diff
@@ -1,5 +1,33 @@
...
+import retrofit2.converter.moshi.MoshiConverterFactory
+
+interface FeedApiService {
+    @GET(FEED_QUERY)
+    fun getFeed(): Single<FeedMetadataDto>
+
+    companion object {
+        private const val FEED_QUERY = "feed.json"
+        const val BASE_URL = "https://assets.swishvideoapp.com/Android/demo/"
+        const val THUMBNAILS_URL = "catalog/thumbnails/"
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/network/FeedApiService.kt]: `THUMBNAILS_URL` isn't just the prefix? I would rename it as `PREFIX_THUMBNAILS_URL`. Also don't we need to append to it the `BASE_URL`. (BASE_URL + PREFIX_THUMBNAILS_URL)

File: app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt [hunk 4]
```diff
@@ -1,5 +1,35 @@
...
+    abstract fun insertAll(feedItems: List<FeedItemEntity>)
+
+    @Delete
+    abstract fun deleteAsync(feedItem: FeedItemEntity): Completable
+
+    @Query("DELETE FROM items")
+    abstract fun deleteAll()
+
+    @Query("SELECT * FROM items")
+    abstract fun getAll(): Observable<List<FeedItemEntity>>
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt]: For the method that are returning a `RxJava` object you suffixed the method with `Async`, let do the same here in order to be consistent with the other method of the class.

File: app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt [hunk 5]
```diff
@@ -1,5 +1,35 @@
...
+
+    @Query("SELECT count(*) FROM items")
+    abstract fun count(): Single<Int>
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt]: Same comment.

File: app/src/main/java/com/lightricks/feedexercise/database/Entities.kt [hunk 6]
```diff
@@ -1,5 +1,14 @@
 package com.lightricks.feedexercise.database
 
-/**
- * todo: add Room's Entity data class(es) here
- */
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity(tableName = "items")
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/Entities.kt]: `items` is too general, please rename the table name to something more descriptive.

File: app/src/main/java/com/lightricks/feedexercise/database/Entities.kt [hunk 7]
```diff
@@ -1,5 +1,14 @@
...
+data class FeedItemEntity(
+    val configuration: String,
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/Entities.kt]: Please declare the fields names explicitly by using the `ColumnInfo` annotation, so that if in the future someone will rename some fields, the database will not be affected by this change.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 8]
```diff
@@ -1,13 +1,145 @@
...
+    private fun makeMetadataDtoFromJson(): FeedMetadataDto? {
+        val moshi: Moshi = Moshi.Builder()
+            .add(KotlinJsonAdapterFactory())
+            .build()
+        val adapter = moshi.adapter(FeedMetadataDto::class.java)
+        val jsonFile = context.assets.open("get_feed_response.json")
+        var jsonString = ""
+        jsonFile.bufferedReader().forEachLine { jsonString += it }
+        return adapter.fromJson(jsonString)
+    }
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: If not in use so please remove.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 9]
```diff
@@ -1,13 +1,145 @@
...
+
+    @Before
+    fun init() {
+        RxAndroidPlugins.initMainThreadScheduler { Schedulers.trampoline() }
+        createDao()
+        createMockApiService()
+        repositoryToTest = FeedRepository(mockApiService, mockFeedDao)
+    }
+
+    private fun createDao() {
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: `createDao` -> `initMockFeedDao`?

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 10]
```diff
@@ -1,13 +1,145 @@
...
+        `when`(mockFeedDao.getAll())
+            .thenReturn(Observable.just(listOfDtoItems.toFeedEntities()))
+    }
+
+    private fun createMockApiService() {
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: `createMockApiService` -> `initMockApiService`?

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt (hunk 9 above)
Comment (nissimLightricks) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: That is because you are not creating an object. A method with the name create should instantiate an object an return it.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 11]
```diff
@@ -1,13 +1,145 @@
...
+        val adapter = moshi.adapter(FeedMetadataDto::class.java)
+        val jsonFile = context.assets.open("get_feed_response.json")
+        var jsonString = ""
+        jsonFile.bufferedReader().forEachLine { jsonString += it }
+        return adapter.fromJson(jsonString)
+    }
+
+    private fun createMetadataDto(listOfItems: List<FeedItemDto>): FeedMetadataDto {
+        return FeedMetadataDto(listOfItems)
+    }
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:58]: Like this method. This method makes a creation.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 12]
```diff
@@ -1,13 +1,145 @@
...
+                templateName = "1",
+                templateThumbnailURI = ""
+            )
+        )
+    }
+
+    @Test
+    fun whenInitCompilesWithoutError_passTest() {
+        assert(true)
+    }
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: I guess you forget to implement the test. Also the assert method from the kotlin package only throws an exception in the case the runtime assertions have been enabled on the JVM using the -ea JVM option. So use a different method.

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt [hunk 13]
```diff
@@ -1,9 +1,57 @@
...
+class FeedRepository(
+    private val feedApiService: FeedApiService,
+    private val feedDao: FeedDao
+):Repository
+{
+    override fun getFeed(): Observable<List<FeedItem>> {
+        return feedDao.getAll()
+            .subscribeOn(Schedulers.io())
+            .map { it.toFeedItems() }
+            .onErrorReturn { emptyList() }
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: We should not swallow the exception.

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt [hunk 14]
```diff
@@ -1,9 +1,57 @@
...
  */
-class FeedRepository {
-    //todo: implement
-}
\ No newline at end of file
+class FeedRepository(
+    private val feedApiService: FeedApiService,
+    private val feedDao: FeedDao
+):Repository
+{
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: Move to the previous line.

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt [hunk 15]
```diff
@@ -1,9 +1,57 @@
...
  * where the data actually comes from (network, database or somewhere else).
  */
-class FeedRepository {
-    //todo: implement
-}
\ No newline at end of file
+class FeedRepository(
+    private val feedApiService: FeedApiService,
+    private val feedDao: FeedDao
+):Repository
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: Add a space after the parentheses and the column. https://kotlinlang.org/docs/coding-conventions.html#colon

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt [hunk 16]
```diff
@@ -1,9 +1,57 @@
...
+        return feedApiService.getFeed()
+            .subscribeOn(Schedulers.io())
+            .flatMapCompletable { feedResponse ->
+                feedDao.refreshDatabaseAsync(feedResponse.templatesMetadata.toFeedEntities())
+            }
+    }
+}
+
+fun List<FeedItemEntity>.toFeedItems(): List<FeedItem> {
+    return map {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: We should give a name to the parameter of the lambda if the lambda is longer than one line.

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt [hunk 17]
```diff
@@ -1,9 +1,57 @@
...
+        FeedItem(
+            id = it.id,
+            thumbnailUrl = FeedApiService.BASE_URL + FeedApiService.THUMBNAILS_URL + it.templateThumbnailURI,
+            isPremium = it.isPremium
+        )
+    }
+}
+
+fun List<FeedItemDto>.toFeedEntities(): List<FeedItemEntity> {
+    return map {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: Same comment.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 18]
```diff
@@ -1,13 +1,145 @@
...
+
+    @Test
+    fun whenRepositoryRefresh_databaseEmpty_willFillDatabaseWithItemsOnTestJson() {
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: The name of the method does not follow our naming convention for test method. `MethodName_StateUnderTest_ExpectedBehavior`

Also I don't understand the state under test, why it is matter that the DB is empty (also I don't see you check/require this state on the test)?

So there it should be something like:
`refresh_shouldFillDBWithItemsFetchFromNetwork` 
I didn't add a special state because it should work no matter the current state it is.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 19]
```diff
@@ -1,13 +1,145 @@
...
+        repositoryToTest.refresh()
+            .test()
+            .await()
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: Why do we need this operator? If you changed the schedulers with the trampoline scheduler you don't have to await.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt [hunk 20]
```diff
@@ -1,13 +1,145 @@
...
+            .assertComplete()
+            .assertNoErrors()
+
+        verify(mockApiService).getFeed()
+        verify(mockFeedDao).refreshDatabaseAsync(listOfDtoItems.toFeedEntities())
+    }
+
+    @Test
+    fun whenGetFeed_databaseExists_willReturnListFromDatabase() {
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: Same comment about the name of the method.

Comment (nissimLightricks) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: getFeed_whileDbNotEmpty_shouldReturnCachedItem

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt [hunk 21]
```diff
@@ -2,41 +2,69 @@ package com.lightricks.feedexercise.ui.feed
...
+        observeFeedItems()
     }
 
-    fun getIsEmpty(): LiveData<Boolean> {
-        //todo: fix the implementation
-        return MutableLiveData()
+    fun refresh() {
+        updateState { copy(isLoading = true) }
+
+        if (!getState().isRefreshing) {
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt]: `isLoading` and `isRefreshing` are taking track of the same operation so remove either one of the two.

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt [hunk 22]
```diff
@@ -25,18 +29,32 @@ class FeedFragment : Fragment() {
     private lateinit var dataBinding: FeedFragmentBinding
     private lateinit var viewModel: FeedViewModel
     private lateinit var feedAdapter: FeedAdapter
+    private lateinit var feedRepository: FeedRepository
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt]: I don't see a reason for holding the repository instance here. 

What's the reason the fragment is keeping a reference to the repository?

File: app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt [hunk 23]
```diff
@@ -25,18 +29,32 @@ class FeedFragment : Fragment() {
...
+            FeedDatabase::class.java,
+            "database"
+        ).build()
+        val roomDao = roomDB.feedDao()
+        return FeedRepository(FeedApiService.feedApiService, roomDao)
+    }
+
     private fun setupViewModel() {
-        viewModel = ViewModelProvider(this, FeedViewModelFactory())
+        viewModel = ViewModelProvider(this, FeedViewModelFactory { feedRepository })
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedFragment.kt]: You should rather call `createFeedRepository`.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedViewModelTest.kt [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+                isPremium = false,
+                thumbnailUrl = "aaa.png"
+            ),
+        )
+    }
+
+    @Test
+    fun whenInitCompilesWithoutError_passTest() {
+        assert(true)
+    }
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedViewModelTest.kt]: Please remove or complete the test.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedViewModelTest.kt [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+
+    @Test
+    fun whenRefreshCalledInViewModel_thenRefreshIsCalledInRepository() {
+        viewModelTest.refresh()
+        //happens once for the init of the view model
+        verify(mockFeedRepository, times(2)).refresh()
+    }
+
```
Comment (nissimLightricks) [inline app/src/androidTest/java/com/lightricks/feedexercise/data/FeedViewModelTest.kt]: naming convention for the method names

**Review by nissimLightricks — CHANGES_REQUESTED** [review]
Excellent job. Round.

File: app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt (hunk 13 above)
Comment (amonsonego) [reply app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt]: We did talk about how  since I assume on my view model that this function cannot fail that I should deal with it here. So i thought it would be nice to just sent an empty list whenever there is an error.

File: app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt (hunk 19 above)
Comment (amonsonego) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: I did change the schedulers to trampoline but still the test fails without the await operator.

Comment (amonsonego) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: more specifically it fails on the assertComplete operator. It passes without it.

Comment (nissimLightricks) [reply app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt]: Try with `RxJavaPlugins.setIoSchedulerHandler { Schedulers.trampoline() }`

File: app/src/main/java/com/lightricks/feedexercise/database/Entities.kt [hunk 26]
```diff
@@ -1,14 +1,15 @@
...
+    @ColumnInfo val configuration: String,
     @PrimaryKey val id: String,
-    val isNew: Boolean,
-    val isPremium: Boolean,
-    val templateName: String,
-    val templateThumbnailURI: String
+    @ColumnInfo val isNew: Boolean,
+    @ColumnInfo val isPremium: Boolean,
+    @ColumnInfo val templateName: String,
+    @ColumnInfo val templateThumbnailURI: String
```
Comment (nissimLightricks) [inline app/src/main/java/com/lightricks/feedexercise/database/Entities.kt]: What about giving a name to the id?

**Review by nissimLightricks — CHANGES_REQUESTED** [review]
Good job. F&S.

**Review by nissimLightricks — APPROVED** [review]
